[2단계 - 자동차 경주 리팩터링] 정(정진우) 미션 제출합니다.

bugoverdose opened the pull request:
안녕하세요.

MVC 패턴으로 리팩토링하는 것이 2단계 미션의 내용인데, 이미 1단계에서 MVC 패턴을 구현해버린 것 같아서 어떻게 할지 고민을 좀 하다가 몇 가지 기능을 추가 및 수정해봤습니다.

1. 게임 재시작 기능 추가 
- 각 게임 종료 후 y 혹은 n을 입력 받아 게임 재실행 여부 결정
- y를 입력받은 경우 자동차명을 받는 과정부터 시작하여 모든 과정 재시작.
- (car 인스턴스들의 위치를 0으로 리셋하고, 동일한 라운드 수로 게임을 다시 시작하는 기능을 구현하려다 말았습니다.)

2. CarRepository 수정 
- findAllCars 메서드로 반환된 cars 배열은 조회 목적으로만 사용되어야 한다는 생각이 들었습니다. cars.clear()처럼 변형을 가할 경우 예외가 발생하도록 Collections.unmodifiableList(cars)를 반환하도록 수정했습니다. 
- clear 메서드 추가: 각 게임이 종료된 후 이전 게임에 사용된 cars를 제거하기 위한 기능. 

3. 모든 util, validator 클래스들의 디폴트 생성자에 private을 추가해봤습니다.

bugoverdose committed:
refactor: add private constructors for all views, utils, constants classes

bugoverdose committed:
docs(README): update feature list

- minor code challenge
- add project structure info

bugoverdose committed:
feat(Game): implement clearCars method

- add test cases
- implement and use clear method at carRepository

bugoverdose committed:
feat(GameController): implement and use endGame method

- clear carRepository and game instance

bugoverdose committed:
fix(CarRepository): prevent direct modification on cars list

- findAllCars returns an unmodifiable list of cars
- fix test cases

bugoverdose committed:
feat(view): implement requestRestartGameInput method

bugoverdose committed:
feat(ValidatorUtils): implement validateAndParseBoolean method

- add test cases

bugoverdose committed:
feat(GameController): implement restart game functionality

- restart on y input
- rename previous run method to playGame method

ddu0422 commented on src/main/java/racingcar/controller/GameController.java:
```diff
@@ -16,9 +17,19 @@
         initGame();
         playGameUntilEnd();
         printGameResult();
+        endGame();
```
GameContoller 내부에서 Input과 Output을 진행하고 있군요 !_!
현재 Controller와 View 부분이 엮여있는 구조로 보이는군요..!

어느 부분에서 View를 나타내야하고 Controller와 View는 어떤 방식으로 소통하면 좋을 지 고민해보면 좋을 것 같아요 :)

bugoverdose replied:
조금 실험적이게 바꿔보았습니다.

ddu0422 commented on src/main/java/racingcar/util/RandomUtils.java:
```diff
@@ -4,6 +4,9 @@
 
     private static final int MAX_NUM_EXCLUDE = 10;
 
+    private RandomUtils() {
+    }
```
사용하지 않는 생성자 private으로 변경하셨군요 !! 💯 💯

ddu0422 commented on src/main/java/racingcar/util/ValidatorUtils.java:
```diff
@@ -44,4 +49,15 @@ public static void validateNoDuplicateCar(List<Car> cars) {
+    }
+
+    private static void validateBooleanString(String value) {
+        if (!value.equals("y") && !value.equals("n")) {
```
현재 value가 null일 순 없겠지만 value가 null이라면 NPE가 발생할 가능성이 있군요 !_!

ddu0422 replied:
상수로 분리해보는 건 어떨까요?_?
1단계에서 진행한 내용은 2단계에서도 동일하게 가져가야하지 않을까요?

ddu0422 commented on src/main/java/racingcar/util/ValidatorUtils.java:
```diff
@@ -44,4 +49,15 @@ public static void validateNoDuplicateCar(List<Car> cars) {
+    }
+
+    private static void validateBooleanString(String value) {
+        if (!value.equals("y") && !value.equals("n")) {
```
y, n 뿐만 아니라 대문자 Y, N도 받을 수 있다면 더 유연하지 않을까요?

bugoverdose replied:
넵 추가해볼께요

ddu0422 commented on src/main/java/racingcar/controller/GameController.java:
```diff
@@ -51,4 +62,14 @@ private void printGameResult() {
+
+    private void endGame() {
+        game.clearCars();
+        game = null;
```
game.clearCars()로 모든 데이터를 삭제했는데 game을 null로 지정한 이유가 있을까요?

bugoverdose replied:
결과적으로 차이는 없지만 game 인스턴스 자체가 더 이상 필요하지 않은 데이터라는 점을 좀 더 명시해준다고 생각해서 함께 썼던걸로 기억해요.

ddu0422 commented on src/main/java/racingcar/controller/GameController.java:
```diff
@@ -16,9 +17,19 @@
+        while (isRunning) {
+            playGame();
+            isRunning = requestNewGameInput();
+        }
```
do while로 변경한다면 불필요한 변수를 제거할 수 있지 않을까요?
지역변수로 선언하긴했지만 값을 변경하는 방법과 값을 변경하지 않는 부분에서 어떤 차이가 있을까요?

bugoverdose replied:
오 진짜 좋은거 같아요

ddu0422 commented on src/test/java/racingcar/util/ValidatorUtilsTest.java:
```diff
@@ -102,4 +103,23 @@ void validateNoDuplicateCar_errorOnDuplicates() {
+        boolean shouldBeFalse = validateAndParseBoolean("n");
+
+        assertThat(shouldBeTrue).isTrue();
+        assertThat(shouldBeFalse).isFalse();
```
만약 대문자도 받게 된다면 테스트를 해야할 경우가 4개로 늘어나겠군요 !_!

ddu0422 reviewed (changes requested):
안녕하세요 정 :) 2단계도 빠르게 구현해주셨네요 !_!
심화 내용을 만들어 진행까지 하셨군요. ㅎ_ㅎ 💯 💯 
간단한 코멘트 남겨드렸으니 확인 부탁드립니다 :)

bugoverdose committed:
refactor(ValidatorUtils): restart game input values to private constants

bugoverdose committed:
feat(ValidatorUtils): allow caseinsensitive for game restart input

bugoverdose committed:
feat(ValidatorUtils): null value input throws RuntimeException, not NPE

bugoverdose committed:
refactor(GameController): remove unnecessary local variable with do while

bugoverdose committed:
refactor(controller): move run logic to main method, remove state from controller, factory for Game domain

bugoverdose committed:
refactor(OutView): separate controller and view logic

ddu0422 commented on src/main/java/racingcar/controller/GameController.java:
```diff
@@ -1,54 +1,24 @@
-    private void playGameUntilEnd() {
-        OutputView.printRoundResultText();
+    public List<List<Car>> playAndGetRoundResults(Game game) {
+        List<List<Car>> roundResults = new ArrayList<>();
```
List<List<Car>>를 활용하면 가독성 저하와 함께 코드량이 많아질 수 있어요.
[일급 컬렉션](https://tecoble.techcourse.co.kr/post/2020-05-08-First-Class-Collection)도 함께 읽어보아요 :)

ddu0422 commented on src/test/java/racingcar/util/ValidatorUtilsTest.java:
```diff
@@ -102,4 +104,41 @@ void validateNoDuplicateCar_errorOnDuplicates() {
+    void validateAndParseBoolean_errorOnNullInput(String nullOrEmptyValue) {
+        assertThatExceptionOfType(RuntimeException.class)
+                .isThrownBy(() -> validateAndParseBoolean(nullOrEmptyValue));
+    }
```
해당 메서드를 호출할 때 RuntimeException이 2군데에서 발생할 수 있군요 !_!
과연.. 빈 값인 경우 둘 중 어떤 메시지를 포함한 RuntimeException이 발생할까요?

ddu0422 commented on src/main/java/racingcar/util/ValidatorUtils.java:
```diff
@@ -44,4 +51,22 @@ public static void validateNoDuplicateCar(List<Car> cars) {
+
+    public static boolean validateAndParseBoolean(String value) {
+        validateBooleanString(value);
+        return value.equalsIgnoreCase(RESTART_GAME_INPUT);
```
💯 💯

ddu0422 commented on src/main/java/racingcar/Application.java:
```diff
@@ -1,11 +1,26 @@
+        do {
+            String carNamesInput = InputView.requestCarNameInput();
+            String totalRoundsInput = InputView.requestTotalRoundsInput();
+            Game game = Game.byUserInput(carNamesInput, totalRoundsInput);
```
사실 지금 상황에선 Domain이 해당 부분까지 나오는 것도 괜찮다고 생각하지만..!
시간이 조금 남아 심심하실 것 같아서 하나 남겨드립니다 :) 해당 부분이 너무 오래 걸릴 것 같다면 다른 내용만 진행하셔도 괜찮습니다!

OutputView에 Domain 객체를 넘겨줄 순 있지만 Game Domain이 해당 부분 생성하는 방식이 괜찮을 지 고민해보면 어떨까요?

해당 부분은 우선순위가 최하위이기 때문에 나머지 글을 읽어보고 생각해보신 후 시간이 괜찮으시다면 고민해봐도 좋을 것 같아요!!

bugoverdose replied:
제가 제대로 이해한건지는 모르겠는데 이 부분 관련해서 한 번 수정해봤습니다.

ddu0422 reviewed (changes requested):
안녕하세요 정 :) 빠르게 진행해주셨군요 🥇  

해당 미션은 완료 처리해도 괜찮지만 조금 심심하실 것 같아 몇가지 코멘트 남겼어요 :)
부담 갖지 말고 생각해주시면 좋을 것 같아요 ㅎ_ㅎ

bugoverdose committed:
feat(Cars): cars to save deep copy of car list

- CarRepository: save the cars to use for logic
- Cars : save the snapshot of the cars for each round
- add test cases on deep copy

bugoverdose committed:
test(ValidatorUtils): separate null and empty input cases

bugoverdose committed:
feat(GameRepository): implement repository to maintain single game instance

bugoverdose committed:
refactor(GameController): use game data via gamerepository instance

ddu0422 commented on src/main/java/racingcar/domain/Cars.java:
```diff
@@ -0,0 +1,19 @@
+
+    private final List<Car> cars;
+
+    public Cars(List<Car> cars) {
+        this.cars = cars.stream()
+                .map(Car::clone)
+                .collect(Collectors.toList());
+    }
```
오호 clone을 사용하셨군요.. 정말 deepcopy를 유지할 수 있을까요!_!
만약 자동차 하나가 중간에 이동을 하는 로직이 추가되면 어떻게 될까요??

this.car와 car의 상태를 보며 디버깅해보시면 좋을 것 같아요 :)

bugoverdose replied:
곰곰히 생각해보니 오직 현재 상태의 데이터만을 담는게 목표였으니 CarDto를 만드는게 나았을 것 같네요.

ddu0422 replied:
@bugoverdose 오호 좋은 방향이군요 ㅎ_ㅎ 👍🏻  👍🏻

ddu0422 reviewed (approved):
안녕하세요 정 :) 
🎆  🎉 첫 미션 고생하셨습니다 🎉 🎆 

이만 머지해도 될 것 같아서 머지할게요!

하나 코멘트 드린 부분이 있는데 같이 고민해보면 좋을 것 같아요 :)
언제든지 의견 내주시고 궁금한 점은 코멘트가 DM 주세요 !_!
